File: backend/src/application/use-cases/RecordConsentUseCase.ts
import { ConsentRecord } from "../../domain/user/ConsentRecord";
import { assertValidConsentVersion } from "../../domain/user/ConsentPolicy";
import { UserNotFoundError } from "../../domain/user/errors/UserNotFoundError";
import { UserRepository } from "../ports/UserRepository";
import { ConsentRepository } from "../ports/ConsentRepository";
import { IdGenerator } from "../ports/IdGenerator";
import { Clock } from "../ports/Clock";

export interface RecordConsentInput {
  userId: string;
  version: string;
}

export interface ConsentRecordView {
  id: string;
  version: string;
  acceptedAt: Date;
}

export class RecordConsentUseCase {
  constructor(
    private readonly userRepository: UserRepository,
    private readonly consentRepository: ConsentRepository,
    private readonly idGenerator: IdGenerator,
    private readonly clock: Clock,
  ) {}

  async execute(input: RecordConsentInput): Promise<ConsentRecordView> {
    const user = await this.userRepository.findById(input.userId);
    if (!user) {
      throw new UserNotFoundError();
    }

    assertValidConsentVersion(input.version);

    const record = ConsentRecord.give({
      id: this.idGenerator.generate(),
      userId: input.userId,
      version: input.version,
      acceptedAt: this.clock.now(),
    });

    await this.consentRepository.save(record);

    return { id: record.id, version: record.version, acceptedAt: record.acceptedAt };
  }
}

File: backend-java/src/main/java/com/financepulse/engine/application/usecases/user/RecordConsentUseCase.java
package com.financepulse.engine.application.usecases.user;

import com.financepulse.engine.application.ports.ConsentRepository;
import com.financepulse.engine.application.ports.IdGenerator;
import com.financepulse.engine.domain.user.ConsentRecord;
import java.time.Instant;

/** RF-046 (ver ADR-0023): registra um novo aceite — trilha append-only, nunca atualizada ou apagada. */
public class RecordConsentUseCase {

    private final ConsentRepository consentRepository;
    private final IdGenerator idGenerator;

    public RecordConsentUseCase(ConsentRepository consentRepository, IdGenerator idGenerator) {
        this.consentRepository = consentRepository;
        this.idGenerator = idGenerator;
    }

    public Output execute(Input input) {
        ConsentRecord consentRecord = ConsentRecord.create(idGenerator.generate(), input.userId(), input.version());

        consentRepository.save(consentRecord);

        return new Output(consentRecord.getId(), consentRecord.getAcceptedAt());
    }

    public record Input(String userId, String version) {
    }

    public record Output(String consentId, Instant acceptedAt) {
    }
}
